backend/compact-connect/lambdas/provider-data-v1/handlers/ingest.py

```diff
@@ -0,0 +1,128 @@
+                'Item': TypeSerializer().serialize(LicenseRecordSchema().dump({
+                    'providerId': provider_id,
+                    'compact': compact,
+                    'jurisdiction': jurisdiction,
+                    **license_post
+                }))['M']
+            }
+        }
+    ]
+
+    try:
+        provider_data = config.data_client.get_provider(  # pylint: disable=missing-kwoa,unexpected-keyword-arg
+            compact=compact,
+            provider_id=provider_id,
+            detail=True,
+            consistent_read=True
+        )
+        # Get all privilege jurisdictions, directly from privilege records
+        privilege_jurisdictions = {
+            record['jurisdiction']
+            for record in provider_data['items']
+            if record['type'] == 'privilege' and record['status'] == 'active'
+        }
+        # Get all the existing license records, by jurisdiction, to find the best data for the provider
+        licenses = {
+            record['jurisdiction']: record
+            for record in provider_data['items']
+            if record['type'] == 'license'
+        }
+    except CCNotFoundException:
+        privilege_jurisdictions = set()
+        licenses = {}
+
+    # Which license do we use for provider data?
+    # If at least one active: last issued active license
+    # If all inactive: last issued inactive license
+    # Set (or replace) the posted license for its jurisdiction
+    licenses[license_post['jurisdiction']] = license_post
+    best_license = find_best_license(licenses.values())
+    if best_license is license_post:
+        logger.info('Updating provider data', provider_id=provider_id, compact=compact)
+        provider_record = populate_provider_record(
+            provider_id=provider_id,
+            license_post=license_post,
+            privilege_jurisdictions=privilege_jurisdictions
+        )
+        # Update our provider data
+        dynamo_transactions.append(
+            {
+                'Put': {
+                    'TableName': config.provider_table_name,
+                    'Item': provider_record
+                }
+            }
+        )
+
+    # Write the records together as a transaction that succeeds or fails as one, to ensure consistency
+    config.dynamodb_client.transact_write_items(TransactItems=dynamo_transactions)
+
+
+def populate_provider_record(*, provider_id: str, license_post: dict, privilege_jurisdictions: set) -> dict:
+    dynamodb_serializer = TypeSerializer()
+    return dynamodb_serializer.serialize((ProviderRecordSchema().dump({
+        'providerId': provider_id,
+        'compact': license_post['compact'],
+        'licenseJurisdiction': license_post['jurisdiction'],
+        # We can't put an empty string set to DynamoDB, so we'll only add the field if it is not empty
+        **({'privilegeJurisdictions': privilege_jurisdictions} if privilege_jurisdictions else {}),
+        **license_post
+    })))['M']
```

Comment:
DynamoDB uses [type descriptors](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html#HowItWorks.DataTypeDescriptors) for data types in its lower level API. Because we're using a lower-level client for this transaction, we have to serialize the data ourselves (or actually use boto3's `TypeSerializer` ourselves). When you feed `TypeSerializer` a `dict` type, it will serialize it as a `Map` type, like `{'M': {...}}`. We only want the value under the `M` key in this case, so we grab it with `['M']`.